plugins/default/src/Ide/Plugin/Pragmas.hs

```diff
@@ -164,3 +164,22 @@ completion lspFuncs _ide complParams = do
                         _xdata = Nothing
                       }
         _ -> return $ Completions $ List []
+
+-- | Find the first non-blank line before the first of (module name / imports / declarations).
+-- Useful for inserting pragmas.
+endOfModuleHeader :: Maybe ParsedModule -> Maybe T.Text -> Range
+endOfModuleHeader mpm contents =
+    case mpm of 
+      Just pm -> 
+        let mod = unLoc $ pm_parsed_source pm
+            modNameLoc = getLoc <$> hsmodName mod
+            firstImportLoc = getLoc <$> listToMaybe (hsmodImports mod)
+            firstDeclLoc = getLoc <$> listToMaybe (hsmodDecls mod)
+            line = fromMaybe 0 $ firstNonBlankBefore . _line . _start =<< srcSpanToRange =<<
+                modNameLoc <|> firstImportLoc <|> firstDeclLoc
```

Comment:
If the line of one of these is `0`, then we don't even need to use `contents <- LSP.getVirtualFileFunc lsp . toNormalizedUri $ docId ^. J.uri`, so I think loading `contents` (and actual texts) only when we need is better for performance.

Reply:
and by using annotations (for comments), we can optimize this even more, so that most of "non-script" cases we don't need to load `contents`.